# PISA 2012 Assessment
## by Amira Noaman

## Preliminary Wrangling

PISA is a survey of students' skills and knowledge as they approach the end of compulsory education. The aim of this test is to examine how well students have digested the school curriculum and how well prepared they are for life beyond school. Approximately a half million students in 65 economies joined the PISA 2012 assessment of reading, mathematics and science representing about 28 million 15-year-olds globally.

The full 2012 PISA dataset can be found [here.](https://s3.amazonaws.com/udacity-hosted-downloads/ud507/pisa2012.csv.zip)

The PISA data dictionary, which gives a brief explaination of each feature, can be found [here.](https://s3.amazonaws.com/udacity-hosted-downloads/ud507/pisadict2012.csv)

In [7]:
# import all packages and set plots to be embedded inline
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sb

%matplotlib inline

In [8]:
df = pd.read_csv('pisa2012.csv', encoding='latin-1')
df.head()

C:\Users\LENOVO\anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3444: DtypeWarning: Columns (15,16,17,21,22,23,24,25,26,30,31,36,37,45,65,123,155,156,157,158,159,160,161,162,163,164,165,166,167,168,169,170,171,284,285,286,287,288,289,290,291,292,293,294,295,296,297,298,299,300,301,302,303,307,308,309,310,311,312,313,314,315,316,317,318,319,320,321,322,323,324,325,326,327,328,329,330,331,332,333,334,335,336,337,338,339,340,341,342,343,344,345,346,347,348,349,350,351,352,353,354,355,356,357,376,377,378,379,380,381,382,383,384,385,386,387,388,389,390,391,392,393,394,395,396,397,398,399,400,401,402,403,475) have mixed types.Specify dtype option on import or set low_memory=False.
  exec(code_obj, self.user_global_ns, self.user_ns)


Unnamed: 0      CNT  SUBNATIO  STRATUM      OECD       NC  SCHOOLID  \
0           1  Albania     80000  ALB0006  Non-OECD  Albania         1   
1           2  Albania     80000  ALB0006  Non-OECD  Albania         1   
2           3  Albania     80000  ALB0006  Non-OECD  Albania         1   
3           4  Albania     80000  ALB0006  Non-OECD  Albania         1   
4           5  Albania     80000  ALB0006  Non-OECD  Albania         1   

   STIDSTD  ST01Q01  ST02Q01  ...  W_FSTR75  W_FSTR76 W_FSTR77 W_FSTR78  \
0        1       10      1.0  ...   13.7954   13.9235  13.1249  13.1249   
1        2       10      1.0  ...   13.7954   13.9235  13.1249  13.1249   
2        3        9      1.0  ...   12.7307   12.7307  12.7307  12.7307   
3        4        9      1.0  ...   12.7307   12.7307  12.7307  12.7307   
4        5        9      1.0  ...   12.7307   12.7307  12.7307  12.7307   

   W_FSTR79 W_FSTR80 WVARSTRR VAR_UNIT SENWGT_STU  VER_STU  
0    4.3389  13.0829       19        1     0.2098  22NOV13  
1    4.3389  13.0829       19        1     0.2098  22NOV13  
2    4.2436  12.7307       19        1     0.1999  22NOV13  
3    4.2436  12.7307       19        1     0.1999  22NOV13  
4    4.2436  12.7307       19        1     0.1999  22NOV13  

[5 rows x 636 columns]

In [9]:
df.shape

(485490, 636)

### What is the structure of your dataset?
The dataset contains information about 485,490 students and presents 636 features.

### What is/are the main feature(s) of interest in your dataset?

The 636 is a huge amount of features so I decided to select some of them to build my analysis. I'm planning to select:
1. Access to the internet
2. Motivation
3. Wealth
4. Country
5. Gender

### What features in the dataset do you think will help support your investigation into your feature(s) of interest?

I sorted the features here accordance to the most important at the top to the least important at the bottom. I believe that the features at the top will have a significant impact on the academic scores.

## Data Assessment

In [10]:
df.duplicated().any()

False

Now we made sure there are no duplicates so we can deal with the whole data given 

In [60]:
#organize features by category list

#CNT: country, ST04Q01: gender, WEALTH: wealth
student_info=['CNT','ST04Q01','WEALTH']

#PV1MATH: Overall Math Score, #PV1READ: Overall Math Score, #PV1SCIE: Overall Math Score
scores=['PV1MATH','PV1READ','PV1SCIE']


#ST29Q02: Worthwhile for Work, ST29Q05: Worthwhile for Career Chances, ST29Q07: Important for Future Study
#ST29Q08: Helps to Get a Job
motivation=['ST29Q02','ST29Q05','ST29Q07','ST29Q08']

#ST26Q06: Possessions-Internet
internet = ['ST26Q06']


In [19]:
df['ST26Q04'].value_counts()

Yes    411040
No      62837
Name: ST26Q04, dtype: int64

### Student Info: Country, Gender, Wealth

In [28]:
df[student_info].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 485490 entries, 0 to 485489
Data columns (total 3 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   CNT      485490 non-null  object 
 1   ST04Q01  485490 non-null  object 
 2   WEALTH   479597 non-null  float64
dtypes: float64(1), object(2)
memory usage: 11.1+ MB


**For the wealth attribute we have some missing values that we have to deal with**

In [34]:
(df['WEALTH'].isnull().sum()/len(df))*100

1.213825207522297

**Only 1% is missing so no need for extra effort to fix the null values**

In [11]:
df['CNT'].value_counts()

Mexico                      33806
Italy                       31073
Spain                       25313
Canada                      21544
Brazil                      19204
                            ...  
Florida (USA)                1896
Perm(Russian Federation)     1761
Massachusetts (USA)          1723
Connecticut (USA)            1697
Liechtenstein                 293
Name: CNT, Length: 68, dtype: int64

**It was mentioned in the description of the dataset that we have a total of 65 participating countries --> so how come we have 68 countries in the value_counts()**

**There are extra 3 values are not correct and obviously we can see that the last 3 values here are incorrectly labeled.
They are 'Florida (USA)', 'Massachusetts (USA)', 'Connecticut (USA)'. This should be replaced with United States of America**

In [13]:
df['ST04Q01'].value_counts()

Female    245064
Male      240426
Name: ST04Q01, dtype: int64

**Gender is almost evenly split so no need for normalization**

### Scores [Math, Reading, Science]

In [35]:
df[scores].head()

PV1MATH   PV1READ   PV1SCIE
0  406.8469  249.5762  341.7009
1  486.1427  406.2936  548.9929
2  533.2684  401.2100  499.6643
3  412.2215  547.3630  438.6796
4  381.9209  311.7707  361.5628

In [36]:
df[scores].describe()

PV1MATH        PV1READ        PV1SCIE
count  485490.000000  485490.000000  485490.000000
mean      469.621653     472.004640     475.769824
std       103.265391     102.505523     101.464426
min        19.792800       0.083400       2.648300
25%       395.318600     403.600700     404.457300
50%       466.201900     475.455000     475.699400
75%       541.057800     544.502500     547.780700
max       962.229300     904.802600     903.338300

In [37]:
df[scores].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 485490 entries, 0 to 485489
Data columns (total 3 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   PV1MATH  485490 non-null  float64
 1   PV1READ  485490 non-null  float64
 2   PV1SCIE  485490 non-null  float64
dtypes: float64(3)
memory usage: 11.1 MB


**We have no missing values in the 3 provided scores**

### Motivation

In [39]:
df[motivation].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 485490 entries, 0 to 485489
Data columns (total 4 columns):
 #   Column   Non-Null Count   Dtype 
---  ------   --------------   ----- 
 0   ST29Q02  315473 non-null  object
 1   ST29Q05  315231 non-null  object
 2   ST29Q07  315066 non-null  object
 3   ST29Q08  315232 non-null  object
dtypes: object(4)
memory usage: 14.8+ MB


**We need to combine the 4 features related to motivation**

### Internet Possession

In [61]:
df[internet].value_counts()

ST26Q06
Yes        402040
No          71142
dtype: int64

**Almost no missing values here**

### Data Assessment Conclusion

- Delete unneccessary features


- Rename feature labels


- Fix incorrectly labeled countries: Replace values of 'Florida (USA)','Massachusetts (USA)', and 'Connecticut (USA)' with 'United States of America'


- Motivation: Combine the features into one feature by assigning point values and taking average.

## Data Cleaning

**1. Delete unneccessary features by making copy of the dataframe with the chosen features**

In [87]:
features=['CNT', 'ST04Q01', 'WEALTH','PV1MATH','PV1READ','PV1SCIE','ST29Q02','ST29Q05','ST29Q07','ST29Q08','ST26Q06']
df_clean=df[features]

In [88]:
df_clean.head()

CNT ST04Q01  WEALTH   PV1MATH   PV1READ   PV1SCIE         ST29Q02  \
0  Albania  Female   -2.92  406.8469  249.5762  341.7009  Strongly agree   
1  Albania  Female    0.69  486.1427  406.2936  548.9929  Strongly agree   
2  Albania  Female   -0.23  533.2684  401.2100  499.6643  Strongly agree   
3  Albania  Female   -1.17  412.2215  547.3630  438.6796             NaN   
4  Albania  Female   -1.17  381.9209  311.7707  361.5628  Strongly agree   

          ST29Q05         ST29Q07         ST29Q08 ST26Q06  
0           Agree           Agree  Strongly agree      No  
1           Agree        Disagree        Disagree     Yes  
2  Strongly agree  Strongly agree  Strongly agree     Yes  
3             NaN             NaN             NaN     Yes  
4  Strongly agree  Strongly agree  Strongly agree     Yes

**2.Combine motivation features**

**Instrumental Motivation:** worthwhile for work, worthwhile for career changes, important for future study

#ST29Q02: Worthwhile for Work, ST29Q05: Worthwhile for Career Chances, ST29Q07: Important for Future Study
#ST29Q08: Helps to Get a Job

In [89]:
scale={'Strongly agree':4,'Agree': 3,'Disagree': 2, 'Strongly disagree':1}

In [90]:
# We will assign point values to each feature category and combine into one feature
for feature in motivation:
    df_clean.loc[:,feature]=df_clean.loc[:,feature].map(scale) #use the scale dictionary to map values to each answer for each feature
df_clean['motivation']=df_clean.loc[:,motivation].sum(axis=1, skipna=False)/len(motivation) # Combine for each student the 4 answers' values as one value

C:\Users\LENOVO\anaconda3\lib\site-packages\pandas\core\indexing.py:1773: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._setitem_single_column(ilocs[0], value, pi)
C:\Users\LENOVO\AppData\Local\Temp/ipykernel_13408/2096203219.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_clean['motivation']=df_clean.loc[:,motivation].sum(axis=1, skipna=False)/len(motivation) # Combine for each student the 4 answers' values as one value


**3. Delete the extra 4 features related to motivation, we'll just leave the combined column representing all of them**

In [91]:
df_clean.drop(motivation,axis=1,inplace=True)
df_clean.head()

C:\Users\LENOVO\anaconda3\lib\site-packages\pandas\core\frame.py:4906: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().drop(


CNT ST04Q01  WEALTH   PV1MATH   PV1READ   PV1SCIE ST26Q06  motivation
0  Albania  Female   -2.92  406.8469  249.5762  341.7009      No        3.50
1  Albania  Female    0.69  486.1427  406.2936  548.9929     Yes        2.75
2  Albania  Female   -0.23  533.2684  401.2100  499.6643     Yes        4.00
3  Albania  Female   -1.17  412.2215  547.3630  438.6796     Yes         NaN
4  Albania  Female   -1.17  381.9209  311.7707  361.5628     Yes        4.00

**4. Rename undescriptive columns**

In [92]:
names={'CNT':'country', 'ST04Q01':'gender','WEALTH':'wealth_index','PV1MATH':'math_score','PV1READ':'reading_score',
       'PV1SCIE':'science_score', 'ST26Q06':'internet_possession'}

df_clean.rename(columns=names,inplace=True)

C:\Users\LENOVO\anaconda3\lib\site-packages\pandas\core\frame.py:5039: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().rename(


In [93]:
df_clean.head()

country  gender  wealth_index  math_score  reading_score  science_score  \
0  Albania  Female         -2.92    406.8469       249.5762       341.7009   
1  Albania  Female          0.69    486.1427       406.2936       548.9929   
2  Albania  Female         -0.23    533.2684       401.2100       499.6643   
3  Albania  Female         -1.17    412.2215       547.3630       438.6796   
4  Albania  Female         -1.17    381.9209       311.7707       361.5628   

  internet_possession  motivation  
0                  No        3.50  
1                 Yes        2.75  
2                 Yes        4.00  
3                 Yes         NaN  
4                 Yes        4.00

**5. Replace values of 'Florida (USA)','Massachusetts (USA)', and 'Connecticut (USA)' with 'United States of America'**

In [94]:
df_clean.country.value_counts()[['United States of America',
                                 'Florida (USA)','Massachusetts (USA)',
                                 'Connecticut (USA)' ]]

United States of America    4978
Florida (USA)               1896
Massachusetts (USA)         1723
Connecticut (USA)           1697
Name: country, dtype: int64

In [95]:
cities=['Florida (USA)','Massachusetts (USA)','Connecticut (USA)']

df_clean['country'].replace(cities, 'United States of America', inplace=True)

C:\Users\LENOVO\anaconda3\lib\site-packages\pandas\core\generic.py:6619: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return self._update_inplace(result)


In [96]:
df_clean.country.value_counts()

Mexico                      33806
Italy                       31073
Spain                       25313
Canada                      21544
Brazil                      19204
                            ...  
Latvia                       4306
New Zealand                  4291
Iceland                      3508
Perm(Russian Federation)     1761
Liechtenstein                 293
Name: country, Length: 65, dtype: int64

**Now we correctly have 65 countries**

## Dataset is Ready!

In [98]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 485490 entries, 0 to 485489
Data columns (total 8 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   country              485490 non-null  object 
 1   gender               485490 non-null  object 
 2   wealth_index         479597 non-null  float64
 3   math_score           485490 non-null  float64
 4   reading_score        485490 non-null  float64
 5   science_score        485490 non-null  float64
 6   internet_possession  473182 non-null  object 
 7   motivation           312694 non-null  float64
dtypes: float64(5), object(3)
memory usage: 29.6+ MB


In [97]:
df_clean.head()

country  gender  wealth_index  math_score  reading_score  science_score  \
0  Albania  Female         -2.92    406.8469       249.5762       341.7009   
1  Albania  Female          0.69    486.1427       406.2936       548.9929   
2  Albania  Female         -0.23    533.2684       401.2100       499.6643   
3  Albania  Female         -1.17    412.2215       547.3630       438.6796   
4  Albania  Female         -1.17    381.9209       311.7707       361.5628   

  internet_possession  motivation  
0                  No        3.50  
1                 Yes        2.75  
2                 Yes        4.00  
3                 Yes         NaN  
4                 Yes        4.00

In [102]:
df_clean.to_csv('pisa2012_clean.csv', index_label=False)